packages/server/src/workers/reindex.ts

```diff
@@ -115,56 +140,36 @@ export async function execReindexJob(job: Job<ReindexJobData>): Promise<void> {
- * @param searchRequest - The search filter to find resources to reindex.
  * @param job - The current job.
+ * @returns The result of reindexing the next page of results.
  */
-async function processPage(searchRequest: SearchRequest, job: Job<ReindexJobData>): Promise<void> {
-  const { resourceTypes, currentTimestamp, count } = job.data;
-  const systemRepo = getSystemRepo();
-  const { hasMore, newCount, nextTimestamp } = await systemRepo.withTransaction(async (conn) => {
-    const bundle = await systemRepo.search(searchRequest);
-    let newCount = count ?? 0;
-    let nextTimestamp = currentTimestamp;
-
-    if (bundle.entry?.length) {
-      const resources: Resource[] = [];
-      // Since the page size could be relatively large (1k+), preferring a simple for loop
-      // eslint-disable-next-line @typescript-eslint/prefer-for-of
-      for (let i = 0; i < bundle.entry.length; i++) {
-        const resource = bundle.entry[i].resource as Resource;
-        resources.push(resource);
-        nextTimestamp = resource.meta?.lastUpdated as string;
-      }
-      await systemRepo.reindexResources(conn, ...resources);
-      newCount += resources.length;
-    }
-
-    const hasMore = bundle.link?.some((link) => link.relation === 'next') ?? false;
-    return { hasMore, newCount, nextTimestamp };
-  });
-
+async function processPage(job: Job<ReindexJobData>): Promise<ReindexResult> {
+  const { resourceTypes, count, currentTimestamp } = job.data;
   const resourceType = resourceTypes[0];
-  if (hasMore) {
-    job.data.results[resourceType] = nextTimestamp;
 
-    // Log progress and update status in the AsyncJob resource periodically
-    if (Math.floor(newCount / progressLogThreshold) !== Math.floor((newCount - batchSize) / progressLogThreshold)) {
-      globalLogger.info('Reindex in progress', {
-        resourceType,
-        latestJobId: job.id,
-        nextTimestamp,
-        count,
-        elapsedTime: `${Date.now() - job.data.startTime} ms`,
-      });
-      await updateStatus(job);
-    }
+  const searchRequest = searchRequestForNextPage(job);
+  let newCount = count ?? 0;
+  let nextTimestamp: string | undefined;
+  try {
+    const systemRepo = getSystemRepo();
+    await systemRepo.withTransaction(async (conn) => {
+      const bundle = await systemRepo.search(searchRequest);
+      if (bundle.entry?.length) {
+        const resources = bundle.entry.map((e) => e.resource as Resource);
+        await systemRepo.reindexResources(conn, resources);
+        newCount += resources.length;
+      }
 
-    // Enqueue job to handle next page of the current resource type
-    await addReindexJobData({
-      ...job.data,
-      currentTimestamp: nextTimestamp,
-      count: newCount,
+      if (bundle.link?.some((link) => link.relation === 'next')) {
+        nextTimestamp = bundle.entry?.[bundle.entry?.length - 1].resource?.meta?.lastUpdated;
```

Comment:
If we switch to `cursor`, you could instead use `nextCursor = parseSearchRequest(url).cursor`